src/modules/musig/main_impl.h

```diff
@@ -449,196 +476,195 @@ int secp256k1_musig_partial_signature_parse(const secp256k1_context* ctx, secp25
-    secp256k1_fe_normalize(&noncep.x);
-    secp256k1_fe_get_b32(&sig64[0], &noncep.x);
-    secp256k1_scalar_get_b32(&sig64[32], &s);
-
+    secp256k1_scalar_clear(&k[0]);
+    secp256k1_scalar_clear(&k[1]);
     return 1;
 }
 
-int secp256k1_musig_partial_sig_verify(const secp256k1_context* ctx, const secp256k1_musig_session *session, const secp256k1_musig_session_signer_data *signer, const secp256k1_musig_partial_signature *partial_sig, const secp256k1_xonly_pubkey *pubkey) {
-    unsigned char msghash[32];
-    secp256k1_scalar s;
-    secp256k1_scalar e;
-    secp256k1_scalar mu;
+int secp256k1_musig_partial_sig_verify(const secp256k1_context* ctx, const secp256k1_musig_partial_signature *partial_sig, const unsigned char *pubnonce66, const secp256k1_xonly_pubkey *pubkey, const secp256k1_musig_pre_session *pre_session, const secp256k1_musig_session_cache *session_cache) {
+    secp256k1_scalar mu, b, e, s;
     secp256k1_gej pkj;
+    secp256k1_ge nonces[2];
     secp256k1_gej rj;
+    secp256k1_gej tmp;
     secp256k1_ge pkp;
-    secp256k1_ge rp;
     int overflow;
+    int i;
+    int combined_nonce_parity;
 
     VERIFY_CHECK(ctx != NULL);
     ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
-    ARG_CHECK(session != NULL);
-    ARG_CHECK(signer != NULL);
     ARG_CHECK(partial_sig != NULL);
+    ARG_CHECK(pubnonce66 != NULL);
     ARG_CHECK(pubkey != NULL);
-    ARG_CHECK(session->magic == session_magic);
-    ARG_CHECK(session->round == 2);
-    ARG_CHECK(signer->present);
+    ARG_CHECK(pre_session != NULL);
+    ARG_CHECK(pre_session->magic == pre_session_magic);
+    ARG_CHECK(session_cache != NULL);
 
-    secp256k1_scalar_set_b32(&s, partial_sig->data, &overflow);
-    if (overflow) {
-        return 0;
+    secp256k1_musig_session_cache_load(&b, &e, &combined_nonce_parity, session_cache);
+
+    /* Compute "effective" nonce rj = nonces[0] + b*nonces[1] */
+    /* TODO: use multiexp */
+    for (i = 0; i < 2; i++) {
+        secp256k1_pubkey n;
+        int ret;
+        if (!secp256k1_ec_pubkey_parse(ctx, &n, &pubnonce66[i*33], 33)) {
+            return 0;
+        }
+        ret = secp256k1_pubkey_load(ctx, &nonces[i], &n);
+        /* Successfully parsed pubkey is always valid */
+        VERIFY_CHECK(ret);
     }
-    secp256k1_musig_compute_messagehash(ctx, msghash, session);
-    secp256k1_scalar_set_b32(&e, msghash, NULL);
+    secp256k1_gej_set_ge(&rj, &nonces[1]);
+    secp256k1_ecmult(&ctx->ecmult_ctx, &rj, &rj, &b, NULL);
+    secp256k1_gej_add_ge_var(&rj, &rj, &nonces[0], NULL);
 
+    if (!secp256k1_xonly_pubkey_load(ctx, &pkp, pubkey)) {
+        return 0;
+    }
     /* Multiplying the messagehash by the musig coefficient is equivalent
      * to multiplying the signer's public key by the coefficient, except
      * much easier to do. */
-    secp256k1_musig_coefficient(&mu, session->pre_session.pk_hash, signer->index);
+    secp256k1_musig_coefficient(&mu, pre_session, &pkp.x);
```

Comment:
`pkp.x` not normalized?

Reply:
we just loaded `pkp`, so it's coordinates are normalized